### PR TITLE
Bootstrap the bridge

### PR DESCRIPTION
**Pull request by abbyck**
1. Add email inbound support - using `smtp-server`
2. Generate AS registration file for HS
3. Parse the from address and set the userid accordingly

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+const Cli = require("matrix-appservice-bridge").Cli;
+const AppServiceRegistration = require("matrix-appservice-bridge").AppServiceRegistration;
```
Comment (Half-Shot) [inline app.js:2]:
```suggestion
const { AppServiceRegistration } = require("matrix-appservice-bridge");
```

File: app.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+const Cli = require("matrix-appservice-bridge").Cli;
```
Comment (Half-Shot) [inline app.js:1]:
```suggestion
const { Cli } = require("matrix-appservice-bridge");
```

File: config/config.sample.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#Sample config
```
Comment (Half-Shot) [inline config/config.sample.yaml:1]:
```suggestion
# Sample config
```

File: config/config.sample.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+bridge:
+  # Domain part of the bridge, e.g. matrix.org
+  domain: "localhost"
+  # This should be your publicly-facing URL because Discord may use it to
```
Comment (Half-Shot) [inline config/config.sample.yaml:5]: Discord?

File: src/bridge.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            },
+
+            onEvent: function(request, context) {
+                // events from matrix
+                const event = request.getData();
+                console.log(event)
+                if (event.type !== "m.room.message" || !event.content) {
+                    return;
+                }
+                console.log(`${event.sender}: ${event.content.body}`)
```
Comment (Half-Shot) [inline src/bridge.js:22]: Thought: Might be worth trying to use a logging system for this (in the future). The bridge library bundles a logger, https://github.com/matrix-org/matrix-appservice-bridge/#logging.

File: src/bridge.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+const Bridge = require('matrix-appservice-bridge').Bridge;
+const {startSMTP}= require('./email');
```
Comment (Half-Shot) [inline src/bridge.js:2]: Might be worth getting a linter rule for consistent spacing in brackets.

```suggestion
const { startSMTP }= require('./email');
```

File: src/email.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+const SMTPServer = require("smtp-server").SMTPServer;
+const MailParser = require('mailparser').MailParser;
```
Comment (Half-Shot) [inline src/email.js:2]:
```suggestion
const { SMTPServer } = require("smtp-server");
const { MailParser } = require('mailparser');
```

File: src/email.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+module.exports.startSMTP = new SMTPServer({
+    secure: false,
+    logger: false,
+    disabledCommands: ['AUTH', 'STARTTLS'],
+    authOptional: true,
+
+    onData: function (stream, session, callback) {
+        let subject, text;
+        let mailparser = new MailParser();
+        let fromAdd = addrs.parseOneAddress(session.envelope.mailFrom.address);
```
Comment (Half-Shot) [inline src/email.js:17]: I think these can be const?

```suggestion
        const mailparser = new MailParser();
        const fromAdd = addrs.parseOneAddress(session.envelope.mailFrom.address);
```

File: src/email.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        // console.log(session.envelope.mailFrom.address)
+
+        mailparser.on('headers', headers => {
+            subject = headers.get('subject');
```
Comment (Half-Shot) [inline src/email.js:21]: Subject is currently not used?

File: src/email.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        });
+
+        mailparser.on('data', data => {
+            if (data.type === 'text') {
+                text = data.text;
+            }
+        });
+
+        mailparser.on('end', () => {
+            console.log(text);
```
Comment (Half-Shot) [inline src/email.js:31]: Would be best to validate that text was set before sending here, in case we never got any data?

File: src/email.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        let mailparser = new MailParser();
+        let fromAdd = addrs.parseOneAddress(session.envelope.mailFrom.address);
+        // console.log(session.envelope.mailFrom.address)
+
+        mailparser.on('headers', headers => {
+            subject = headers.get('subject');
+        });
+
+        mailparser.on('data', data => {
+            if (data.type === 'text') {
```
Comment (Half-Shot) [inline src/email.js:25]: I'm curious if it's possible to have multiple `data` events before the `end`, as you sometimes see with large emails and streams where you get chunks. In that case, should we append to text?

File: src/email.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+                text = data.text;
+            }
```
Comment (Half-Shot) [inline src/email.js:27]: Do we want to log a warning if the type was not text?

**Review by Half-Shot — COMMENTED** [review]
This is looking promising. I've added a mixture of linting comments, and ideas inline.

File: src/email.js (hunk 9 above)
Comment (abbyck) [reply src/email.js:21]: What's the best way to use subject in this context?

**Pull request merged**